query-planner-js/src/buildPlan.ts

```diff
@@ -4027,6 +4104,81 @@ function operationForEntitiesFetch(
   return new Operation(subgraphSchema, 'query', entitiesCall, variableDefinitions, operationName);
 }
 
+class Finder {
+  constructor(readonly target: FieldDefinition<any>) {}
+
+  fieldName(): string {
+    return this.target.name;
+  }
+
+  entityType(): NamedType {
+    const entityType = this.target.type;
+    assert(entityType, 'finder must return a defined type');
+    return entityType as NamedType;
+  }
+
+  arguments(): Readonly<ArgumentDefinition<FieldDefinition<any>>[]> {
+    return this.target.arguments();
+  }
+
+  argument(idx: number): { inputType: InputType, name: string } {
+    const arg = this.target.arguments()[idx];
+    assert(arg, `Argument at index ${idx} exists`);
+    assert(arg.type, 'type exists on argument');
+    return {
+      inputType: arg.type,
+      name: arg.name,
+    };
+  }
+}
+
+function operationForFindersFetch({
+  subgraphSchema,
+  selectionSet,
+  allVariableDefinitions,
+  operationName,
+  finder,
+}: {
+  subgraphSchema: Schema,
+  selectionSet: SelectionSet,
+  allVariableDefinitions: VariableDefinitions,
+  operationName?: string,
+  finder: Finder,
+}): Operation {
+  const inputField = finder.argument(0);
+  const variableDefinitions = new VariableDefinitions();
+  const variable = new Variable(inputField.name);
+  const metadata = federationMetadata(subgraphSchema);
+  assert(metadata, 'Expected schema to be a federation subgraph');
+  variableDefinitions.add(new VariableDefinition(subgraphSchema, variable, inputField.inputType));
```

Comment:
We should be checking for, and deal with clashes between the name of the variable we use for the key, and the other variables from the user query as those can absolutely clash. Fwiw, I think that's where the `variableName` of `FetchInput` could be useful: we cannot assume that use the name of the key as name of the variable will always work and we need to deal with that.

I'll note that for `_entities`, there is code in the query plan execution that validates that no user variables are named "representations" and that's how we avoid clash (it's not ideal as it's a bit of an arbitrary limitation on users, but it's at least handled gracefully and no-one has complained about it that I know of, so fine so far in practice). I don't think we can use that approach here: rejecting a variable name at execution because some key used in the query somewhere has the name is imo too limiting, and could even leak information it shouldn't if that key is not public.